neural_lam/datastore/base.py

```diff
@@ -0,0 +1,371 @@
+    `weather_dataset.WeatherDataset` class (which inherits from
+    `torch.utils.data.Dataset` and uses the datastore to access the data).
+
+    # Forecast vs analysis data
+    If the datastore is used represent forecast rather than analysis data, then
+    the `is_forecast` attribute should be set to True, and returned data from
+    `get_dataarray` is assumed to have `analysis_time` and `forecast_time`
+    dimensions (rather than just `time`).
+
+    # Ensemble vs deterministic data
+    If the datastore is used to represent ensemble data, then the `is_ensemble`
+    attribute should be set to True, and returned data from `get_dataarray` is
+    assumed to have an `ensemble_member` dimension.
+
+    """
+
+    is_ensemble: bool = False
+    is_forecast: bool = False
+
+    @property
+    @abc.abstractmethod
+    def root_path(self) -> Path:
+        """
+        The root path to the datastore. It is relative to this that any derived
+        files (for example the graph components) are stored.
+
+        Returns
+        -------
+        pathlib.Path
+            The root path to the datastore.
+
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def config(self) -> collections.abc.Mapping:
+        """The configuration of the datastore.
+
+        Returns
+        -------
+        collections.abc.Mapping
+            The configuration of the datastore, any dict like object can be
+            returned.
+
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def step_length(self) -> int:
+        """The step length of the dataset in hours.
+
+        Returns:
+            int: The step length in hours.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_vars_units(self, category: str) -> List[str]:
+        """Get the units of the variables in the given category.
+
+        Parameters
+        ----------
+        category : str
+            The category of the variables.
+
+        Returns
+        -------
+        List[str]
+            The units of the variables.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_vars_names(self, category: str) -> List[str]:
+        """Get the names of the variables in the given category.
+
+        Parameters
+        ----------
+        category : str
+            The category of the variables.
+
+        Returns
+        -------
+        List[str]
+            The names of the variables.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_num_data_vars(self, category: str) -> int:
+        """Get the number of data variables in the given category.
+
+        Parameters
+        ----------
+        category : str
+            The category of the variables.
+
+        Returns
+        -------
+        int
+            The number of data variables.
+
+        """
+        pass
+
+    @abc.abstractmethod
+    def get_normalization_dataarray(self, category: str) -> xr.Dataset:
```

Comment:
The name and documentation of this should use `standardize` rather than `normalize`, as this is about parameters for rescaling to mean 0 and $\sigma=1$.

Reply:
Down the line we probably want to allow for both somehow (normalizing some features to $[0,1]$), but that does not have to be solved in this PR.

Reply:
Ah, you're right. That is important distinction. I guess in general this would be called a _scaling_ of variables, and we're applying a scaling that is _standardizing_ rather than _normalizing_

Reply:
I did this refactoring in https://github.com/leifdenby/neural-lam/commit/5263d2ced70b5cca93f042dd9497e662756671b7